File: is_valid_subsequence/is_valid_subsequence.go
package main

import (
	"fmt"
)

func isValidSubsequence(array []int, sequence []int) bool {
	if len(sequence) > len(array) {
		return false
	}

	seqIndex := 0
	for _, num := range array {
		if seqIndex == len(sequence) {
			break
		}
		if num == sequence[seqIndex] {
			seqIndex++
		}
	}
	return seqIndex == len(sequence)
}

func main() {
	array := []int{5, 1, 22, 15, 6, -1, 8, 10}
	sequence := []int{1, 6, -1, 10}
	fmt.Println(isValidSubsequence(array, sequence))
}

File: is_valid_subsequence/is_valid_subsequence.py
# O(n) time O(1) space
def is_valid_subsequence(array, sequence):
  if len(sequence) > len(array):
    return False
  
  seq_index = 0

  for i in range(len(array)):
    if seq_index == len(sequence):
      break
    if array[i] == sequence[seq_index]:
      seq_index += 1

  return seq_index == len(sequence)

array = [5, 1, 22, 15, 6, -1, 8, 10]
sequence = [1, 6, -1, 10]
print(is_valid_subsequence(array, sequence))
